docs/design/2021-09-13-parser-as-submodule-of-tidb.md

```diff
@@ -0,0 +1,89 @@
+
+## Background
+
+I will explain why and how to migration parser back to the tidb repository.
+
+### Why did we move parser into a separate repository?
+
+The PR  is [tidb#2923: Move TiDB parser to a separate repository](https://github.com/pingcap/tidb/issues/7923). In short, when we migrated to go module 111, @tiancaiamao decided to do so, because
+
+> We can put generated `parser.go` in the repository directly, rather than generate it using Makefile script. The drawback of this way is that every time `parser.y` is touched, there will be many lines change in `parser.go` . Then the TiDB repo will be inflate quickly.
+>
+>A better way is moving parser to a separated repo, so every time parser is changed, only go.mod need to change accordingly.
+
+Personally, that is a weird argument for me. Indeed, pingcap/tidb does not store the diff of `parser.go` anymore, but pingcap/parser does. How is moving the inflation from one repo to another better?
+
+Another argument is about the messy dependency. The separation does improve the case.
+
+### What is wrong with a separate repository?
+
+In fact, there is no big problem. However the development becomes inconvenient. You could argue that we have workarounds, but why not make the development easier?
+
+1. We always need to ensure the version of two go modules. Cherry-picking becomes troublesome since the import version must be a specific branch of pingcap/parser.
+
+2. It is a mysql parser, but is closely related to TiDB(basically everything except ast/mysql packages). You must code TiDB logic into the "standalone mysql parser" first before your real implementation, which means at least two PRs, even for small changes.
+
+3. Before the PR of parser merged, the PR of TiDB must use module directive to redirect a specific version of parser to fix CI.
+
+Problems above have defeated the previous argument `only go.mod need to change accordingly`. We only separated code into two repositories, actual work did not decrease. However, our development experience does getting worse.
+
+## Detailed Plan
+
+Obviously, we could move back to the old good way. But it must be a go sub-module, because:
+
+1. To keep compatibility. Parser has been imported by other tools and other users.
+2. To keep the clean dependency. That is why it is separated. We don't want to make things worse again.
+
+The detailed plan:
+
+1. Create parser directory, a sub-module, in pingcap/tidb, which is synchronized with pingcap/parser, with `replace github.com/pingcap/parser => ./parser`.
+2. Stop sending new PRs and issues to pingcap/parser, new PRs should be directly sent to pingcap/tidb. And merge recent PRs in pingcap/parser as much as possible. Changes will be synchronized from pingcap/parser to pingcap/tidb(manually), of course.
+3. Announce the deprecation of pingcap/parser to possible users. Clean up old issues and PRs, create it in pingcap/tidb again, close it, or just forget about it.
```

Comment:
(issues can be transferred, no need to close and reopen)

Reply:
https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository, this, right? I know...

Reply:
Updated.